gateway-ha/src/main/java/io/trino/gateway/ha/router/ProcessedRequest.java

```diff
@@ -0,0 +1,524 @@
+                }
+            }
+        }
+
+        if (node.getClass() == ShowColumns.class) {
+            tables.add(qualifyName(((ShowColumns) node).getTable()));
+        }
+        if (node.getClass() == Table.class) {
+            tables.add(qualifyName(((Table) node).getName()));
+        }
+        for (Node child : node.getChildren()) {
+            getTables(child, tables);
+        }
+    }
+
+    private QualifiedName qualifyName(QualifiedName table)
+    {
+        List<String> tableParts = table.getParts();
+        return switch (tableParts.size()) {
+            case 1:
+                yield QualifiedName.of(defaultCatalog, defaultSchema, tableParts.getFirst());
+            case 2:
+                yield QualifiedName.of(defaultCatalog, tableParts.getFirst(), tableParts.get(1));
+            case 3:
+            default:
+                yield QualifiedName.of(tableParts.getFirst(), tableParts.get(1), tableParts.get(2));
+        };
+    }
+
+    private Optional<String> extractUserFromBearerAuth(String header, String userField)
+    {
+        log.debug("Trying to extract user from bearer token");
+        int space = header.indexOf(' ');
+        if ((space < 0) || !header.substring(0, space).equalsIgnoreCase("bearer")) {
+            return Optional.empty();
+        }
+
+        String token = header.substring(space + 1).trim();
+        ObjectMapper mapper = new ObjectMapper();
+
+        if (header.split("\\.").length == 3) { //this is probably a JWS
+            log.debug("Trying to extract from JWS");
+            token = header.split("\\.")[1];
+            try {
+                JsonNode node = mapper.readTree(new String(Base64.getDecoder().decode(token)));
+                if (node.has(userField)) {
+                    log.debug("Trying to extract user from JWS json. User: " + node.get(userField).asText());
+                    return Optional.of(node.get(userField).asText());
+                }
+            }
+            catch (JsonProcessingException e) {
+                log.warn("Could not deserialize bearer token as json");
+            }
+        }
+
+        if (isGetUserFromOAuth) {
+            String responseBody = googleUserInfo(token);
+            if (responseBody.contains(userField)) {
+                mapper = new ObjectMapper();
+                try {
+                    JsonNode node = mapper.readTree(responseBody);
+                    if (node.has(userField)) {
+                        log.debug("Trying to extract user from json. User: " + node.get(userField).asText());
+                        return Optional.of(node.get(userField).asText());
+                    }
+                }
+                catch (JsonProcessingException ex) {
+                    log.debug("Could not deserialize token info response to json: " + responseBody);
+                }
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    private String googleUserInfo(String token)
```

Comment:
why googleUserInfo? 
Token can be from any IDP - right?

Reply:
yep - leftover from early development

Reply:
any IDP that supports the userinfo endpoint should work